### PR TITLE
Implement CN-GPU proof-of-work hash

### PR DESCRIPTION
**Pull request by fireice-uk**
@psychocrypt Thanks for 3 weeks of hard work on this one.

### CODE REVIEW

File: src/crypto/pow_hash/cn_slow_hash_hard_intel.cpp [hunk 1]
```diff
@@ -244,7 +241,7 @@ void cn_slow_hash<MEMORY, ITER, VERSION>::explode_scratchpad_hard()
 	x6 = _mm_load_si128(spad.as_xmm() + 10);
 	x7 = _mm_load_si128(spad.as_xmm() + 11);
 
-	for(size_t i = 0; VERSION > 0 && i < 16; i++)
+	for (size_t i = 0; VERSION > 0 && i < 16; i++)
```
Comment (psychocrypt) [inline src/crypto/pow_hash/cn_slow_hash_hard_intel.cpp]: not required code style change

File: src/crypto/pow_hash/cn_slow_hash_hard_intel.cpp [hunk 2]
```diff
@@ -227,8 +224,8 @@ void cn_slow_hash<MEMORY, ITER, VERSION>::implode_scratchpad_hard()
 	_mm_store_si128(spad.as_xmm() + 11, x7);
 }
 
-template <size_t MEMORY, size_t ITER, size_t VERSION>
-void cn_slow_hash<MEMORY, ITER, VERSION>::explode_scratchpad_hard()
+template<size_t MEMORY, size_t ITER, size_t VERSION>
+void cn_slow_hash<MEMORY,ITER,VERSION>::explode_scratchpad_hard()
```
Comment (psychocrypt) [inline src/crypto/pow_hash/cn_slow_hash_hard_intel.cpp]: note: not required code style link (there are a lot of it)

File: src/crypto/pow_hash/cn_slow_hash_soft.cpp [hunk 3]
```diff
@@ -187,8 +182,8 @@ struct aesdata
 
 inline uint32_t sub_word(uint32_t key)
 {
-	return (saes_sbox[key >> 24] << 24) | (saes_sbox[(key >> 16) & 0xff] << 16) |
-		   (saes_sbox[(key >> 8) & 0xff] << 8) | saes_sbox[key & 0xff];
+	return (saes_sbox[key >> 24 ] << 24)   | (saes_sbox[(key >> 16) & 0xff] << 16 ) |
```
Comment (psychocrypt) [inline src/crypto/pow_hash/cn_slow_hash_soft.cpp]: why do you add wild spaces. due to the tab indention it is not possible to align code via spaces.

Comment (fireice-uk) [reply src/crypto/pow_hash/cn_slow_hash_soft.cpp]: I will run it through clang format

File: src/crypto/pow_hash/cn_slow_hash_soft.cpp [hunk 4]
```diff
@@ -439,51 +434,84 @@ void cn_slow_hash<MEMORY, ITER, VERSION>::explode_scratchpad_soft()
...
-	uint64_t adbc_carry = adbc < ad ? 1 : 0;
-
-	// multiplier * multiplicand = product_hi * 2^64 + product_lo
-	uint64_t product_lo = bd + (adbc << 32);
-	uint64_t product_lo_carry = product_lo < bd ? 1 : 0;
-	*product_hi = ac + (adbc >> 32) + (adbc_carry << 32) + product_lo_carry;
-
-	return product_lo;
+  // multiplier   = ab = a * 2^32 + b
+  // multiplicand = cd = c * 2^32 + d
```
Comment (psychocrypt) [inline src/crypto/pow_hash/cn_slow_hash_soft.cpp]: note: space instead of tabs

File: src/crypto/pow_hash/cn_slow_hash_soft.cpp [hunk 5]
```diff
@@ -518,11 +546,11 @@ void cn_slow_hash<MEMORY, ITER, VERSION>::software_hash(const void *in, size_t l
 		idx = scratchpad_ptr(ax.v64x0);
 		if(VERSION > 0)
 		{
-			int64_t n = idx.as_qword(0);
-			int32_t d = idx.as_dword(2);
+			int64_t n  = idx.as_qword(0);
```
Comment (psychocrypt) [inline src/crypto/pow_hash/cn_slow_hash_soft.cpp]: one space to much before the assign.

**Comment by fireice-uk** [comment]
Please merge despite clang errors, we will get that fixed in the post.

**Pull request merged**